.azure-pipelines/impacted_area_testing/get_test_scripts.py

```diff
@@ -0,0 +1,108 @@
+
+    topo_type = match.group()
+    if topo_type in ['mgmttor', 'm0', 'mc0', 'mx']:
+        # certain testbed types are in 't0' category with different names.
+        topo_type = 't0'
+    if topo_type in ['multidut-tgen']:
+        topo_type = 'tgen'
+    return topo_type
+
+
+def collect_scripts_by_topology_type(features, location):
+    """
+    This function collects all test scripts under the impacted area and category them by topology type.
+
+    Args:
+        Features: The impacted area defined by features
+        Location: The location of test scripts
+
+    Returns:
+        Dict: A dict of test scripts categorized by topology type.
+    """
+    # Recursively find all files starting with "test_" and ending with ".py"
+    # Note: The full path and name of files are stored in a list named "files"
+    scripts = []
+    for feature in features.split(","):
+        feature_path = os.path.join(location, feature)
+        for root, dirs, script in os.walk(feature_path):
+            for s in script:
+                if s.startswith("test_") and s.endswith(".py"):
+                    scripts.append(os.path.join(root, s))
+    scripts = natsorted(scripts)
+
+    # Open each file and search for regex pattern
+    pattern = re.compile(r"[^@]pytest\.mark\.topology\(([^\)]*)\)")
+
+    # Init the dict to record the mapping of topology type and test scripts
+    test_scripts_per_topology_type = {}
+    for topology_type in PR_TOPOLOGY_TYPE:
+        test_scripts_per_topology_type[topology_type] = []
+
+    for s in scripts:
+        # Remove prefix from file name:
+        script_name = s[len(location) + 1:]
+        if script_name in EXCLUDE_TEST_SCRIPTS:
+            continue
+
+        try:
+            with open(s, 'r') as script:
+                for line in script:
+                    # Get topology type of script from mark `pytest.mark.topology`
+                    match = pattern.search(line)
+                    if match:
+                        for topology in match.group(1).split(","):
+                            topology_mark = topology.strip().strip('"').strip("'")
+                            if topology_mark == "any":
+                                for key in ["t0", "t1"]:
+                                    if script_name not in test_scripts_per_topology_type[key]:
+                                        test_scripts_per_topology_type[key].append(script_name)
+                            else:
+                                topology_type = topo_name_to_type(topology_mark)
+                                if topology_type in test_scripts_per_topology_type \
+                                        and script_name not in test_scripts_per_topology_type[topology_type]:
+                                    test_scripts_per_topology_type[topology_type].append(script_name)
+        except Exception as e:
+            raise Exception('Exception occurred while trying to get topology in {}, error {}'.format(s, e))
```

Comment:
Suggest extracting a method to improve the readability of this deep for-loop.

Reply:
There is two loop here, the outer one is to traverse each line in script to get the topology mark. 
The inner one is to traverse each topology in topology mark. 
Is it really necessary to use another function?

Reply:
Not necessary, but improves code readability.

Reply:
Add a new function `distribute_scripts_to_PR_checkers`